File: bindings/java-jni/src/main/java/org/maplibre/nativejni/map/TileLodMode.java
package org.maplibre.nativejni.map;

/** Tile level-of-detail algorithm used by map tile options. */
public final class TileLodMode {
  public static final TileLodMode DEFAULT = new TileLodMode(0, "DEFAULT");
  public static final TileLodMode DISTANCE = new TileLodMode(1, "DISTANCE");

  private final int nativeValue;
  private final String name;

  private TileLodMode(int nativeValue, String name) {
    this.nativeValue = nativeValue;
    this.name = name;
  }

  public int nativeValue() {
    if (name == null) {
      throw new IllegalArgumentException("Unknown tile LOD mode cannot be used as an input");
    }
    return nativeValue;
  }

  public int rawValue() {
    return nativeValue;
  }

  public static TileLodMode fromNative(int nativeValue) {
    return switch (nativeValue) {
      case 0 -> DEFAULT;
      case 1 -> DISTANCE;
      default -> new TileLodMode(nativeValue, null);
    };
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof TileLodMode value && nativeValue == value.nativeValue;
  }

  @Override
  public int hashCode() {
    return Integer.hashCode(nativeValue);
  }

  @Override
  public String toString() {
    return name != null ? name : "TileLodMode(" + nativeValue + ")";
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/map/TileLodMode.kt
package org.maplibre.nativeffi.map

/** Tile level-of-detail algorithm used by map tile options. */
public enum class TileLodMode(public val nativeValue: Int) {
  DEFAULT(0),
  DISTANCE(1),
  UNKNOWN(-1);

  public companion object {
    internal fun fromNative(nativeValue: UInt): TileLodMode = fromNative(nativeValue.toInt())

    public fun fromNative(nativeValue: Int): TileLodMode =
      entries.firstOrNull { it.nativeValue == nativeValue } ?: UNKNOWN
  }
}
